mongodb/node-mongodb-native
# refactor(NODE-5419): move csfle source and tests into the driver

### Description

#### What is changing?

This PR depends on https://github.com/mongodb/libmongocrypt/pull/667.

This is a large PR but there's not much that can be done about that.  I'll highlight the relevant changes here.

- FLE code has been moved from `encryption` into the `src`.  The FLE logic was basically ported directly as-is into `src`, with three modifications
-- we no longer inject `mongodb` into any of the files.  Instead, each file is a top-level module.
-- we no longer inject BSON into any FLE logic.  Each FLE class requires from `bson` directly now.  And the _bson properties have been removed from each class as well.
-- I converted the imports to `import` syntax instead of `require` syntax.  I did this while debugging, it wasn't necessary, but once it was done I figured I'd leave it because it's one fewer thing to adjust when I convert the files to TS (up next).  I can revert this now though, if y'all would like.
- the tests have been ported into `integration/node-specific` and `unit/client-side-encryption`.
-- the unit tests were mostly left as-is, I plan to update them more when they're converted to TS.  I did have to fix imports though.  I also deleted some redundant tests.
-- any tests that required a "live server" have been moved into integration.
-- shared library tests were moved into `integration`, because we have integration testing environments that have the shared library installed.

Some other misc changes

- obviously, the driver has been updated to consume AutoEncrypter directly from `src`.  changes were pretty minimal here
- I moved the FLE import logic into `deps.ts` to align with our existing optional deps pattern.
- the shared library tests now spawn subprocesses, since loading the shared library is a process-wide change and the tests were interfering with each other.
- a new filter was added, GenericPredicateFilter, that abstracts away the need to manually skip tests in before each hooks by running a specified predicate function in the filter.

**NOTE FOR REVIEWERS**

Until an alpha is release of mongodb-client-encryption containing the changes to the FLE bindings, pretty much all CI is red.  The important tests here are the custom dependency tests with a pinned commit, which pass.  

All the integration tests and unit tests pass locally against a local copy the changes in https://github.com/mongodb/libmongocrypt/pull/667 too.  Once we merge the bindings changes, I'll release an alpha, pull the alpha into this PR and rerun CI.

To test the changes:

- make sure you've checked out PR#667 from libmongocrypt.
- make sure your environment is set up for FLE tests (kms servers, credentials, ect)
- build and link the bindings: `FLE_BINDINGS_PATH=<> ./etc/tooling/fle.sh link --build-libmongocrypt`
- run the tests normally


##### Is there new documentation needed for these changes?

Not yet.

#### What is the motivation for this change?






### Release Highlight



### Fill in title or leave empty for no highlight



### Double check the following

- [ ] Ran `npm run check:lint` script
  - Example: `feat(NODE-1234)!: rewriting everything in coffeescript`
